docs/flow/config-language/expressions/types_and_values.md

```diff
@@ -6,3 +6,145 @@ weight: 100
 ---
 
 # Types and values
+
+## Types
+
+River uses the following types for its values:
+
+* `number`: Any numeric value, like `3` or `3.14`.
+* `string`: A sequence of Unicode characters representing text, like `"Hello, world!"`.
+* `bool`: A boolean value, either `true` or `false`.
+* `array`: A sequence of values, like `[1, 2, 3]`. Elements within the
+  list are indexed by whole numbers, starting with zero.
+* `object`: A group of values which are identified by named labels, like
+  `{ name = "John" }`.
+* `function`: A value representing a routine which can be executed with
+  arguments to compute another value, like `env("HOME")`. Functions take zero
+  or more arguments as input and always return a single value as output.
+* `null`: A type that has no value.
+
+### Special Types
+
+#### Secrets
+
+A `secret` is a special type of string which is never displayed to the user.
+`string` values may be assigned to an attribute expecting a `secret`, but never
+the inverse; it is not possible to convert a secret to a string or assign a
+secret to an attribute expecting a string.
+
+Secrets cannot be constructed by users, but are returned by components which
+deal with sensitive values.
+
+#### Capsules
+
+River has a special type called a `capsule`, which represents an internal type
+used by Grafana Agent for building telemetry pipelines. Capsule values can
+never be constructed by the user, but can be used in expressions and assigned
+to attributes. Capsules are represented to the user as
+`capsule("SOME_INTENRAL_NAME")`. An attribute expecting a capsule can only be
+given a capsule of the same internal type. The specific type of capsule
+expected is explicitly documented for any component which uses or exports them.
+
+
+The `metrics.remote_write` component below exports an attribute called
+`forwarder`, which is a `capsule("metrics.Receiver")` type. This can then be
+used in the `forward_to` attribute of `metrics.scrape`, which expects a list of
+`capsule("metrics.Receiver")`:
+
+```river
+metrics.remote_write "default" {
+  remote_write {
+    url = "http://localhost:9090/api/v1/write"
+  }
+}
+metrics.scrape "default" {
+  targets = [/* ... */]
+  // Have scraped metrics be sent to metrics.remote_write.default's receiver.
+  forward_to = [metrics.remote_write.default.receiver]
+}
+```
+
+## Numbers
+River handles integers, unsigned integers and floating-point values as a single
+'number' type which simplifies writing and reading River configuration files.
+
+When a value has to be decoded into a Go type, River will perform the necessary
+conversion. The language also supports defining numbers using the scientific e
+notation.
```

Comment:
nit: we shouldn't be talking about implementation details since those for change. I think it's fair to drop this sentence entirely for now.